File: packages/platform/src/ipc/customer.ts
import { invoke } from "@tauri-apps/api/core";
import type {
  CustomerDtoProfile,
  CustomerIpcCreateRequest,
  CustomerIpcCreateResponse,
  CustomerIpcGetRequest,
  CustomerIpcGetResponse,
  CustomerIpcListRequest,
  CustomerIpcListResponse,
  CustomerIpcUpdateRequest,
  CustomerIpcUpdateResponse,
} from "@desk/contracts";

export type CustomerProfile = CustomerDtoProfile;

/**
 * List customers with optional search and pagination.
 *
 * @author Xiaoman
 * @created 2026-07-21
 */
export async function customerList(
  input: CustomerIpcListRequest = {},
): Promise<{ items: CustomerProfile[]; total: number }> {
  const response = await invoke<CustomerIpcListResponse>("customer_list", { request: input });
  try {
    const parsed = JSON.parse(response.customers_json ?? "[]") as CustomerProfile[];
    return {
      items: Array.isArray(parsed) ? parsed : [],
      total: response.total ?? 0,
    };
  } catch {
    return { items: [], total: 0 };
  }
}

/**
 * Fetch one customer profile by id.
 *
 * @author Xiaoman
 * @created 2026-07-21
 */
export async function customerGet(id: string): Promise<CustomerProfile> {
  const request: CustomerIpcGetRequest = { id };
  const response = await invoke<CustomerIpcGetResponse>("customer_get", { request });
  return JSON.parse(response.profile_json) as CustomerProfile;
}

/**
 * Create a new customer profile.
 *
 * @author Xiaoman
 * @created 2026-07-21
 */
export async function customerCreate(
  input: CustomerIpcCreateRequest,
): Promise<CustomerProfile> {
  const response = await invoke<CustomerIpcCreateResponse>("customer_create", { request: input });
  return JSON.parse(response.profile_json) as CustomerProfile;
}

/**
 * Update an existing customer profile.
 *
 * @author Xiaoman
 * @created 2026-07-21
 */
export async function customerUpdate(
  input: CustomerIpcUpdateRequest,
): Promise<CustomerProfile> {
  const response = await invoke<CustomerIpcUpdateResponse>("customer_update", { request: input });
  return JSON.parse(response.profile_json) as CustomerProfile;
}

File: crates/app/src/commands/customer.rs
//! Customer profile Tauri IPC commands.
//!
//! 作者：Xiaoman
//! 创建时间：2026-07-21

use common::contracts::{
    CustomerIpcCreateRequest, CustomerIpcCreateResponse, CustomerIpcGetRequest,
    CustomerIpcGetResponse, CustomerIpcListRequest, CustomerIpcListResponse,
    CustomerIpcUpdateRequest, CustomerIpcUpdateResponse,
};
use customer::app::{CreateCustomer, GetCustomer, ListCustomers, UpdateCustomer};

use crate::state::AppState;

/// List customers with optional search and pagination.
///
/// 作者：Xiaoman
/// 创建时间：2026-07-21
#[tauri::command]
pub async fn customer_list(
    state: tauri::State<'_, AppState>,
    request: CustomerIpcListRequest,
) -> Result<CustomerIpcListResponse, String> {
    let store = state.customer_store.clone();
    tauri::async_runtime::spawn_blocking(move || ListCustomers::execute(store.as_ref(), request))
        .await
        .map_err(|error| error.to_string())?
}

/// Fetch one customer profile by id.
///
/// 作者：Xiaoman
/// 创建时间：2026-07-21
#[tauri::command]
pub async fn customer_get(
    state: tauri::State<'_, AppState>,
    request: CustomerIpcGetRequest,
) -> Result<CustomerIpcGetResponse, String> {
    let store = state.customer_store.clone();
    tauri::async_runtime::spawn_blocking(move || GetCustomer::execute(store.as_ref(), request))
        .await
        .map_err(|error| error.to_string())?
}

/// Create a new customer profile.
///
/// 作者：Xiaoman
/// 创建时间：2026-07-21
#[tauri::command]
pub async fn customer_create(
    state: tauri::State<'_, AppState>,
    request: CustomerIpcCreateRequest,
) -> Result<CustomerIpcCreateResponse, String> {
    let store = state.customer_store.clone();
    tauri::async_runtime::spawn_blocking(move || CreateCustomer::execute(store.as_ref(), request))
        .await
        .map_err(|error| error.to_string())?
}

/// Update an existing customer profile.
///
/// 作者：Xiaoman
/// 创建时间：2026-07-21
#[tauri::command]
pub async fn customer_update(
    state: tauri::State<'_, AppState>,
    request: CustomerIpcUpdateRequest,
) -> Result<CustomerIpcUpdateResponse, String> {
    let store = state.customer_store.clone();
    tauri::async_runtime::spawn_blocking(move || UpdateCustomer::execute(store.as_ref(), request))
        .await
        .map_err(|error| error.to_string())?
}
